cpp/src/Slice/Parser.cpp

```diff
@@ -953,6 +953,19 @@ Slice::Contained::isTopLevel() const
     return dynamic_pointer_cast<Unit>(container()) != nullptr;
 }
 
+ModulePtr
+Slice::Contained::getTopLevelModule() const
+{
+    if (auto parent = dynamic_pointer_cast<Contained>(container()))
+    {
+        return parent->getTopLevelModule();
+    }
+    // `Module` has its own implementation of this function. So reaching here means we hit an
+    // element which is a top-level non-module. The parser will report an error for it, but
+    // until we exit (at the end of parsing) this element will exist, and needs to be handled.
+    return nullptr;
```

Comment:
Can this really happen?

Reply:
Yeah for sure. You could compile this Slice file:
```
#pragma once
class MyClass {} // a top-level class!
```
Of course, the compiler will report an error `"classes can only be defined within a module"` during the parsing phase, and we won't generate any code (since an error occurred), but we still happily continue parsing until we reach the end of the file.

Only the code-gen (and some validation) can assume that this is non-null. Because the error will of prevented us from hitting those phases otherwise. But the scanning and parsing phases can make no such assumption.